apis/apiextensions/v1/composition_types.go

```diff
@@ -265,18 +275,97 @@ type PatchPolicy struct {
 
+// A Combine configures a patch that combines more than
+// one input field into a single output field.
+type Combine struct {
+	// Variables are the list of variables whose values will be retrieved and
+	// combined.
+	// +kubebuilder:validation:MinItems=1
+	Variables []CombineVariable `json:"variables"`
+
+	// Strategy defines the strategy to use to combine the input variable values.
+	// Currently only String is supported.
+	// +kubebuilder:validation:Enum=String
+	Strategy CombineStrategy `json:"strategy"`
+
+	// String declares that input variables should be combined into a single
+	// string, using the relevant settings for formatting purposes.
+	// +optional
+	String *StringCombine `json:"string,omitempty"`
+}
+
+// Combine calls the appropriate combiner.
+func (c *Combine) Combine(vars []interface{}) (interface{}, error) {
+	var combiner interface {
+		Combine(vars []interface{}) (interface{}, error)
+	}
+
+	switch c.Strategy {
+	case CombineStrategyString:
+		combiner = c.String
+	default:
+		return nil, errors.Errorf(errFmtCombineStrategyNotSupported, string(c.Strategy))
+	}
+
+	// Check for nil interface requires reflection.
+	if reflect.ValueOf(combiner).IsNil() {
+		return nil, errors.Errorf(errFmtCombineConfigMissing, string(c.Strategy))
+	}
+	out, err := combiner.Combine(vars)
+	// Note: There are currently no tests or triggers to exercise this error as
+	// our only strategy ("String") uses fmt.Sprintf, which cannot return an error.
+	return out, errors.Wrapf(err, errFmtCombineStrategyFailed, string(c.Strategy))
+}
+
+// A CombineVariable defines the source of a value that is combined with
+// others to form and patch an output value. Currently, this only supports
+// retrieving values from a field path.
+type CombineVariable struct {
+	// FromFieldPath is the path of the field on the source whose value is
+	// to be used as input.
+	FromFieldPath string `json:"fromFieldPath,omitempty"`
+}
+
+// A CombineStrategy determines what strategy will be applied to combine
+// variables.
+type CombineStrategy string
+
+// CombineStrategy strategy definitions.
+const (
+	CombineStrategyString CombineStrategy = "String"
+	// Additional:
+	// CombineStrategyGoTemplate CombineStrategy = "GoTemplate"
+)
+
+// A StringCombine combines multiple input values into a single string.
+type StringCombine struct {
+	// Format the input using a Go format string. See
+	// https://golang.org/pkg/fmt/ for details.
+	Format string `json:"fmt"`
+}
+
+// Combine returns a single output by running a string format
+// with all of its' input variables.
+func (s *StringCombine) Combine(vars []interface{}) (interface{}, error) {
+	return fmt.Sprintf(s.Format, vars...), nil
+}
+
 // Apply executes a patching operation between the from and to resources.
 // Applies all patch types unless an 'only' filter is supplied.
-func (c *Patch) Apply(from, to runtime.Object, only ...PatchType) error {
+func (c *Patch) Apply(cp, cd runtime.Object, only ...PatchType) error {
```

Comment:
Renamed these fields for clarity as `to` and `from` make no sense in this context. Went with the same names they're passed with by the Composed reconciler.